Implementing ADOPT in Optax

tinker495 opened the pull request:
I recently came across a promising new optimizer at the following link:

[https://github.com/iShohei220/adopt](https://github.com/iShohei220/adopt)

I found it interesting and decided to implement it in Optax. I believe it is working correctly, but I am unsure about the next steps required for a proper contribution.

Specifically, I would appreciate guidance on:

How to add documentation for this new optimizer.
How to write and integrate tests.
The general process for merging this contribution into the Optax repository.
Thank you for your time and assistance. I look forward to your feedback.

iShohei220 commented:
@tinker495 @vroulet 
Hi! Thank you for your cool implementation on our work!
I've read the code and left some comments to make it identical to our PyTorch implementation [here](https://github.com/iShohei220/adopt).

vroulet commented:
Thanks @iShohei220 ! We can't see your comments it seems, maybe you forgot to publish them?

iShohei220 commented:
> Thanks @iShohei220 ! We can't see your comments it seems, maybe you forgot to publish them?

Oops! I'm sorry. I've made them public now.

tinker495 commented:
@vroulet Following your comments, I have moved the code to `contrib`, added tests, and included it in the documentation.
@iShohei220 I've also addressed the points you commented on.

Could you please review the changes?

tinker495 commented on optax/_src/transform.py:
```diff
@@ -307,6 +307,73 @@ def update_fn(updates, state, params=None):
+      )
+    else:
+      mu_ = mu  
+    updates = mu_
```
When using:

`updates = jnp.where(state.count > 0, mu_, otu.tree_zeros_like(mu_, dtype=mu_dtype))`

I found that issues occur in pytest if the structure contains types like list.
I am changing it to use jax.tree_util.tree_map like this:

`updates = jax.tree_util.tree_map(
    lambda m: jnp.where(state.count > 0, m, jnp.zeros_like(m, dtype=mu_dtype)),
    mu_
)`

This modification allows it to handle all types within the structure.

iShohei220 replied:
Seems great! Thank you.

tinker495 commented on optax/_src/transform.py:
```diff
@@ -307,6 +307,73 @@ def update_fn(updates, state, params=None):
+      mu_updates = jax.tree.map(lambda ud, nu: jnp.clip(ud / jnp.maximum(jnp.sqrt(nu), eps), -clip_value, clip_value), updates, state.nu)
+      b1_ = b1
+    else:
+      mu_updates = jax.tree.map(lambda ud, nu: ud / jnp.maximum(jnp.sqrt(nu), eps), updates, nu)
```
For now, I have made the change as you suggested.

However, if it involves `state.nu`, `mu_updates` could become a value close to infinity on the first step.

If you check this comment and the situation is confirmed, I would like to change it back to the previous version. Would that be alright?

tinker495 replied:
![image](https://github.com/user-attachments/assets/1d2fc5b3-666b-44f0-8deb-a5845380f83c)

Also, looking at the content of your paper, it seems that `nu` (which is `nu_t`) is correct, rather than `state.nu` which corresponds to `nu_(t-1)`.

tinker495 replied:
Or how about using below?
```    
b2_ = jnp.where(state.count > 0, b2, 0)
b1_ = jnp.where(state.count > 0, b1, 1)
nu = otu.tree_update_moment_per_elem_norm(updates, state.nu, b2_, 2)
if use_clipping:
  clip_value = clip_value_fn(state.count)
  mu_updates = jax.tree.map(lambda ud, nu: jnp.clip(ud / jnp.maximum(jnp.sqrt(nu), eps), -clip_value, clip_value), updates, state.nu)
else:
  mu_updates = jax.tree.map(lambda ud, nu: ud / jnp.maximum(jnp.sqrt(nu), eps), updates, state.nu)
mu = otu.tree_update_moment(mu_updates, state.mu, b1_, 1)
count_inc = numerics.safe_increment(state.count)
if nesterov:
  mu_ = jax.tree.map(
      lambda m, g: b1 * m + (1 - b1) * g,
      mu,
      mu_updates,
  )
else:
  mu_ = mu
updates = mu_
mu = otu.tree_cast(mu, mu_dtype)
return updates, transform.ScaleByAdamState(count=count_inc, mu=mu, nu=nu)
```

If b1_ becomes 1 on the first step, mu_(t=0) will unconditionally become 0. Also, it seems cleaner than before.

iShohei220 replied:
> Also, looking at the content of your paper, it seems that nu (which is nu_t) is correct, rather than state.nu which corresponds to nu_(t-1).

In Algorithm 1 in the paper, the update is about `mu_(t+1)`, not `mu_t`, so when updating `mu_t`, you should use `nu_(t-1)`.
See Algorithm 3 in the appendix, which is another equivalent representation of Algorithm 1.

<img width="906" alt="Screenshot 2025-04-21 at 12 03 23" src="https://github.com/user-attachments/assets/a7c3f6cb-b786-4390-99db-19b5601e05ee" />

iShohei220 replied:
> However, if using state.nu, mu_updates could become a value close to infinity on the first step.

It's true that, when `nu_0` is close to 0, `mu_updates` tends to be very large.
This sometimes happens in practice and leads to unstable training.
That is why we introduce clipping in the update of `mu` (see the last paragraph of Section 4 in the paper).

tinker495 replied:
> It's true that, when nu_0 is close to 0, mu_updates tends to be very large in the first step.

However, in the current implementation using Optax, `nu_0` is not close to 0, but literally 0.

Therefore, when considering the first step as t=0, the purpose of the code below is to set `b2_ = 0` and `b1_ = 1`, thereby making `nu_0 = g * g` and `mu_0 = 0`.

> Or how about using below?
> 
> b2_ = jnp.where(state.count > 0, b2, 0)
> b1_ = jnp.where(state.count > 0, b1, 1)
> nu = otu.tree_update_moment_per_elem_norm(updates, state.nu, b2_, 2)
> if use_clipping:
>   clip_value = clip_value_fn(state.count)
>   mu_updates = jax.tree.map(lambda ud, nu: jnp.clip(ud / jnp.maximum(jnp.sqrt(nu), eps), -clip_value, clip_value), updates, state.nu)
> else:
>   mu_updates = jax.tree.map(lambda ud, nu: ud / jnp.maximum(jnp.sqrt(nu), eps), updates, state.nu)
> mu = otu.tree_update_moment(mu_updates, state.mu, b1_, 1)
> count_inc = numerics.safe_increment(state.count)
> if nesterov:
>   mu_ = jax.tree.map(
>       lambda m, g: b1 * m + (1 - b1) * g,
>       mu,
>       mu_updates,
>   )
> else:
>   mu_ = mu
> updates = mu_
> mu = otu.tree_cast(mu, mu_dtype)
> return updates, transform.ScaleByAdamState(count=count_inc, mu=mu, nu=nu)
> If b1_ becomes 1 on the first step, mu_(t=0) will unconditionally become 0. Also, the jnp.where call within the 'updates' is removed, which I believe makes the code cleaner than before.

iShohei220 replied:
Ah, I understand your point now.
Yeah, then this one seems reasonable.

```
b2_ = jnp.where(state.count > 0, b2, 0)
b1_ = jnp.where(state.count > 0, b1, 1)
nu = otu.tree_update_moment_per_elem_norm(updates, state.nu, b2_, 2)
if use_clipping:
  clip_value = clip_value_fn(state.count)
  mu_updates = jax.tree.map(lambda ud, nu: jnp.clip(ud / jnp.maximum(jnp.sqrt(nu), eps), -clip_value, clip_value), updates, state.nu)
else:
  mu_updates = jax.tree.map(lambda ud, nu: ud / jnp.maximum(jnp.sqrt(nu), eps), updates, state.nu)
mu = otu.tree_update_moment(mu_updates, state.mu, b1_, 1)
count_inc = numerics.safe_increment(state.count)
if nesterov:
  mu_ = jax.tree.map(
      lambda m, g: b1 * m + (1 - b1) * g,
      mu,
      mu_updates,
  )
else:
  mu_ = mu
updates = mu_
mu = otu.tree_cast(mu, mu_dtype)
return updates, transform.ScaleByAdamState(count=count_inc, mu=mu, nu=nu)
```

tinker495 replied:
So, would it be okay to change it this way?

tinker495 replied:
Okay, then I will make the change to this approach, including the part below:
> `updates = jax.tree_util.tree_map( lambda m: jnp.where(state.count > 0, m, jnp.zeros_like(m, dtype=mu_dtype)), mu_ )`

iShohei220 replied:
Thank you!
I've checked the updated version, and confirmed that there are no issues.

vroulet commented on docs/api/contrib.rst:
```diff
@@ -130,3 +131,8 @@ Sophia
+ADOPT
+~~~~~~
+.. autofunction:: adopt
+.. autofunction:: scale_by_acprop
```
scale_by_adopt

tinker495 replied:
Oops, it seems like automated correction messed this up.

vroulet commented on optax/contrib/_adopt.py:
```diff
@@ -0,0 +1,170 @@
+  zero at initialization. This can happen when parameters (e.g., the last layer of
+  a neural network) are initialized with zeros.
+
+  Let :math:`\alpha_t` represent the learning rate and :math:`\beta_1, \beta_2`,
```
Remove the mathematical introductions in this paragraph and the one and below

tinker495 replied:
Made the changes as you commented.

vroulet commented on optax/contrib/_adopt.py:
```diff
@@ -0,0 +1,170 @@
+  r"""ADOPT: Modified Adam optimizer that can converge with any beta2 value.
+
+  ADOPT (Adaptive Optimization with Provable Theoretical guarantees) is a modified
+  version of Adam that ensures convergence with any beta2 value while maintaining the
```
I would replace 
"that ensures convergence with any beta2 while maintaining the optimal convergence rate"
with 
"that may improve the robustness of Adam with respect to the choice of beta2"

Convergence guarantees depend on assumptions that are generally not satisfied in deep learning. Such statements can then be misleading for practitioners.

tinker495 replied:
Made the changes as you commented.

vroulet commented on optax/contrib/_adopt.py:
```diff
@@ -0,0 +1,170 @@
+    use_clipping: bool = True,
+    clip_value_fn: Callable[[jnp.ndarray], jnp.ndarray] = lambda x: x ** 0.25,
+) -> base.GradientTransformationExtraArgs:
+  r"""ADOPT: Modified Adam optimizer that can converge with any beta2 value.
```
Just put the the acronym here (no description
```
ADOPT (Adaptive Optimization with Provable Theoretical guarantees)
```
and remove it below.

tinker495 replied:
Made the changes as you commented.

vroulet commented on optax/contrib/_adopt.py:
```diff
@@ -0,0 +1,170 @@
+) -> base.GradientTransformation:
+  r"""Rescale updates according to the ADOPT algorithm.
+
+  ADOPT (Modified Adam Can Converge with Any beta2 with the Optimal Rate) is a variant
```
Remove this paragraph. You may simply add a 
```
  .. seealso:: :func:`optax.contrib.adopt`
```
At the end of this function so that you don't duplicate docstrings.

tinker495 replied:
Made the changes as you commented.

vroulet reviewed (commented):
Thank you @tinker495 and @iShohei220 ! Some additional comments below

iShohei220 commented:
I've checked the latest version, and confirmed that the implementation matches the algorithm we described in the paper.
So, there's no problem for me if all the comments by @vroulet are addressed.

tinker495 commented:
@vroulet 
Apologies for the recurring problems with docstrings. I have addressed the points you mentioned.

vroulet commented:
@tinker495 you still need to sing the cla (see failed check [here](https://github.com/google-deepmind/optax/pull/1257/checks?check_run_id=40914983819))

tinker495 commented:
My apologies, I misunderstood the CLA check as reviewer approval.

I have now signed the CLA and made minor corrections to the docstrings.
